File: app/trends_fetcher.py
import time
import csv
from datetime import date
from pathlib import Path

from pytrends.request import TrendReq
from app.config import settings, RAW_DIR


class TrendsFetcher:
    def __init__(self):
        self.geo = settings.TRENDS_GEO
        self.timeframe = settings.TRENDS_TIMEFRAME
        self.max_kw = settings.TRENDS_MAX_KEYWORDS
        self._client = None

    def _get_client(self):
        if self._client is None:
            kwargs = {"hl": "en-US", "tz": 360, "timeout": 8}
            proxy = settings.get_proxy()
            if proxy:
                kwargs["proxies"] = {"https": proxy, "http": proxy}
            self._client = TrendReq(**kwargs)
        return self._client

    def fetch_category_trends(self, category_name, category_id):
        pt = self._get_client()
        results = []

        # Approach 1: daily trending searches
        try:
            df = pt.trending_searches(pn="united_states")
            if df is not None and not df.empty:
                now = date.today().isoformat()
                for i, row in df.head(self.max_kw).iterrows():
                    keyword = str(row.iloc[0]).strip()
                    if not keyword or len(keyword) < 2:
                        continue
                    results.append({
                        "keyword": keyword,
                        "interest_score": max(0, 100 - i),
                        "rank": i + 1,
                        "category": category_name,
                        "date": now,
                    })
                return results
        except Exception:
            pass

        # Approach 2: realtime trending searches
        try:
            df = pt.realtime_trending_searches(pn="US")
            if df is not None and not df.empty:
                now = date.today().isoformat()
                seen = set()
                for _, row in df.iterrows():
                    title = str(row.get("title", "")).strip()
                    if title and len(title) >= 2 and title.lower() not in seen:
                        seen.add(title.lower())
                        results.append({
                            "keyword": title,
                            "interest_score": max(10, 90 - len(results) * 2),
                            "rank": len(results) + 1,
                            "category": category_name,
                            "date": now,
                        })
                    if len(results) >= self.max_kw:
                        break
                if results:
                    return results
        except Exception:
            pass

        # Approach 3: related queries for the category
        try:
            seed_map = {"Business": "business", "Technology": "technology", "Health": "health"}
            seed = seed_map.get(category_name, "news")
            pt.build_payload(
                kw_list=[seed],
                geo=self.geo,
                timeframe=self.timeframe,
                cat=category_id,
            )
            related = pt.related_queries()
            if related and seed in related:
                rising = related[seed].get("rising")
                if rising is not None and not rising.empty:
                    now = date.today().isoformat()
                    for i, (_, row) in enumerate(rising.head(self.max_kw).iterrows()):
                        keyword = str(row.get("query", "")).strip()
                        if not keyword or len(keyword) < 2:
                            continue
                        val = row.get("value", 0)
                        try:
                            score = min(100, int(val)) if val != "Breakout" else 85
                        except (ValueError, TypeError):
                            score = max(0, 80 - i)
                        results.append({
                            "keyword": keyword,
                            "interest_score": score,
                            "rank": i + 1,
                            "category": category_name,
                            "date": now,
                        })
                    if results:
                        return results
        except Exception:
            pass

        if not results:
            raise RuntimeError("所有抓取方式均失败")

        return results

    def fetch_all(self):
        today_str = date.today().isoformat()
        all_results = {}
        for cat_name, cat_id in settings.TRENDS_CATEGORIES.items():
            try:
                items = self.fetch_category_trends(cat_name, cat_id)
                all_results[cat_name] = items
                self._save_csv(cat_name, today_str, items)
                time.sleep(3)
            except Exception as e:
                all_results[cat_name] = []
                print(f"  [WARN] {cat_name} fetch failed: {e}")

        self._save_merged_csv(today_str, all_results)
        return all_results

    def _save_csv(self, category, date_str, items):
        dir_path = Path(RAW_DIR) / date_str
        dir_path.mkdir(parents=True, exist_ok=True)
        filepath = dir_path / f"{category.lower()}.csv"
        with open(filepath, "w", newline="", encoding="utf-8-sig") as f:
            w = csv.DictWriter(f, fieldnames=["keyword", "interest_score", "rank", "category", "date"])
            w.writeheader()
            w.writerows(items)

    def _save_merged_csv(self, date_str, all_results):
        dir_path = Path(RAW_DIR) / date_str
        dir_path.mkdir(parents=True, exist_ok=True)
        filepath = dir_path / "all_merged.csv"
        all_items = []
        for items in all_results.values():
            all_items.extend(items)
        all_items.sort(key=lambda x: x["interest_score"], reverse=True)
        with open(filepath, "w", newline="", encoding="utf-8-sig") as f:
            w = csv.DictWriter(f, fieldnames=["keyword", "interest_score", "rank", "category", "date"])
            w.writeheader()
            w.writerows(all_items)

File: app/trends-fetcher.js
const fs = require('fs');
const path = require('path');
const https = require('https');
const { parseString } = require('xml2js');
const config = require('./config');
const logger = require('./logger');
const TrendsScraper = require('./trends-scraper');

const CATEGORY_DB_IDS = { Business: 1, Technology: 2, Health: 3 };
const RSS_URL = 'https://trends.google.com/trending/rss?geo=US';

class TrendsFetcher {
  constructor() {
    this.geo = config.defaults.TRENDS_GEO;
    this.maxKw = config.defaults.TRENDS_MAX_KEYWORDS;
  }

  async fetchAll() {
    const todayStr = new Date().toISOString().slice(0, 10);
    const allResults = {};

    // Approach 1: Playwright scraper
    logger.info('fetcher', '尝试方式1: Playwright 渲染抓取');
    try {
      const scraper = new TrendsScraper();
      const scraped = await scraper.scrapeAll();
      let hasData = false;
      for (const [catName, items] of Object.entries(scraped)) {
        if (items && items.length > 0) {
          allResults[catName] = items;
          this._saveCsv(catName, todayStr, items);
          logger.info('fetcher', `[${catName}] 方式1成功: ${items.length} 条`);
          hasData = true;
        } else {
          allResults[catName] = [];
          logger.warn('fetcher', `[${catName}] 方式1无数据`);
        }
      }
      if (hasData) {
        this._saveMergedCsv(todayStr, allResults);
        return allResults;
      }
    } catch (e) {
      logger.warn('fetcher', `方式1全失败: ${e.message}`);
    }

    // Approach 2: RSS feed
    for (const catName of Object.keys(CATEGORY_DB_IDS)) {
      try {
        logger.info('fetcher', `[${catName}] 尝试方式2: RSS`);
        const items = await this._fetchRss(catName);
        if (items.length > 0) {
          allResults[catName] = items;
          this._saveCsv(catName, todayStr, items);
          logger.info('fetcher', `[${catName}] 方式2成功: ${items.length} 条`);
        } else {
          allResults[catName] = [];
          logger.warn('fetcher', `[${catName}] 方式2无数据`);
        }
      } catch (e) {
        allResults[catName] = [];
        logger.warn('fetcher', `[${catName}] 方式2失败: ${e.message}`);
      }
    }

    const hasAny = Object.values(allResults).some(a => a.length > 0);
    if (hasAny) {
      this._saveMergedCsv(todayStr, allResults);
      return allResults;
    }

    // Approach 3: empty
    logger.error('fetcher', '所有方式全部失败，返回空数据');
    for (const catName of Object.keys(CATEGORY_DB_IDS)) {
      allResults[catName] = [];
    }
    return allResults;
  }

  _fetchRss(catName) {
    return new Promise((resolve, reject) => {
      https.get(RSS_URL, { timeout: 15000 }, (res) => {
        let xml = '';
        res.on('data', c => xml += c);
        res.on('end', () => {
          parseString(xml, (err, result) => {
            if (err) return reject(err);
            try {
              const entries = result?.rss?.channel?.[0]?.item || [];
              const today = new Date().toISOString().slice(0, 10);
              const items = entries.slice(0, this.maxKw).map((entry, i) => {
                const keyword = (entry.title?.[0] || '').trim();
                const traffic = (entry['ht:approx_traffic']?.[0] || '').replace(/,/g, '');
                const trafficNum = parseInt(traffic, 10) || 0;
                const score = Math.min(100, Math.max(1, Math.round(trafficNum / 500)));
                return {
                  keyword,
                  interest_score: score || Math.max(1, 100 - i * 2),
                  rank: i + 1,
                  category: catName,
                  date: today,
                };
              }).filter(i => i.keyword && i.keyword.length >= 2);
              resolve(items);
            } catch (e) {
              reject(e);
            }
          });
        });
      }).on('error', reject);
    });
  }

  _saveCsv(category, dateStr, items) {
    const dirPath = path.join(config.RAW_DIR, dateStr);
    fs.mkdirSync(dirPath, { recursive: true });
    const filepath = path.join(dirPath, `${category.toLowerCase()}.csv`);
    const header = 'keyword,interest_score,rank,category,date\n';
    const rows = items.map(i =>
      `"${i.keyword.replace(/"/g, '""')}",${i.interest_score},${i.rank},"${i.category}","${i.date}"`
    ).join('\n');
    fs.writeFileSync(filepath, '\ufeff' + header + rows, 'utf-8');
  }

  _saveMergedCsv(dateStr, allResults) {
    const dirPath = path.join(config.RAW_DIR, dateStr);
    fs.mkdirSync(dirPath, { recursive: true });
    const filepath = path.join(dirPath, 'all_merged.csv');
    const allItems = Object.values(allResults).flat();
    allItems.sort((a, b) => b.interest_score - a.interest_score);
    const header = 'keyword,interest_score,rank,category,date\n';
    const rows = allItems.map(i =>
      `"${i.keyword.replace(/"/g, '""')}",${i.interest_score},${i.rank},"${i.category}","${i.date}"`
    ).join('\n');
    fs.writeFileSync(filepath, '\ufeff' + header + rows, 'utf-8');
  }
}

module.exports = TrendsFetcher;
